### Pandas Lab -- Grouping & Merging

Welcome to today's lab!  It will come in two different parts:  

One section will be devoted to using the `groupby` method in order to answer different questions about our data.  

The second portion will be devoted towards combining grouping & merging to create summary statistics -- one of the more important features you can add to a dataset for statistical modeling.  

### Section I - Grouping

In [1]:
import numpy as np
import pandas as pd

In [214]:
pd.options.display.max_rows = 1000

In [2]:
df = pd.read_csv("/Users/PRSmb/OneDrive/General-Assembly/my-1019-repo/ClassMaterial/Unit2/data/restaurants.csv")

In [3]:
df.head()

id  visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24  2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24  2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24  2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24  2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24  2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252108 entries, 0 to 252107
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                252108 non-null  object 
 1   visit_date        252108 non-null  object 
 2   visitors          252108 non-null  int64  
 3   calendar_date     252108 non-null  object 
 4   day_of_week       252108 non-null  object 
 5   holiday           252108 non-null  int64  
 6   genre             252108 non-null  object 
 7   area              252108 non-null  object 
 8   latitude          252108 non-null  float64
 9   longitude         252108 non-null  float64
 10  reserve_visitors  108394 non-null  float64
dtypes: float64(3), int64(2), object(6)
memory usage: 21.2+ MB


In [9]:
df['visit_date'] = pd.to_datetime(df['visit_date'])

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252108 entries, 0 to 252107
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                252108 non-null  object        
 1   visit_date        252108 non-null  datetime64[ns]
 2   visitors          252108 non-null  int64         
 3   calendar_date     252108 non-null  object        
 4   day_of_week       252108 non-null  object        
 5   holiday           252108 non-null  int64         
 6   genre             252108 non-null  object        
 7   area              252108 non-null  object        
 8   latitude          252108 non-null  float64       
 9   longitude         252108 non-null  float64       
 10  reserve_visitors  108394 non-null  float64       
dtypes: datetime64[ns](1), float64(3), int64(2), object(5)
memory usage: 21.2+ MB


**Question 1:** What restaurant had the highest total amount of visitors throughout the dataset?

In [9]:
# your answer here

In [4]:
df.groupby('id')['visitors'].sum().max()

18717

In [5]:
a = df.groupby('id')['visitors'].sum().reset_index()

In [6]:
a[(a['visitors']==18717)]

id  visitors
172  air_399904bdb7685ca0     18717

In [16]:
df.iloc[172]

id                          air_ba937bf13d40fb24
visit_date                            2016-08-09
visitors                                      22
calendar_date                         2016-08-09
day_of_week                              Tuesday
holiday                                        0
genre                                 Dining bar
area                Tōkyō-to Minato-ku Shibakōen
latitude                                 35.6581
longitude                                139.752
reserve_visitors                             NaN
Name: 172, dtype: object

In [ ]:
air_399904bdb7685ca0

In [ ]:
# Proposed solution

In [70]:
visits = df.groupby('id')['visitors'].sum()

In [59]:
df

id
air_00a91d42b08b08d9    6051
air_0164b9927d20bcc3    1378
air_0241aa3964b7f861    3919
air_0328696196e46f18     921
air_034a3d5b40d5b1b1    3722
                        ... 
air_fea5dc9594450608    3969
air_fee8dcf4d619598e    7496
air_fef9ccb3ba0da2f7    2357
air_ffcc2d5087e1b476    4919
air_fff68b929994bfbd    1369
Name: visitors, Length: 829, dtype: int64

In [62]:
df.reset_index

AttributeError: 'function' object has no attribute 'idmax'

**Question 2:** What was the average difference in attendance between holidays & non-holidays for each restaurant?

In [53]:
# your answer here
?pd.Series.diff

In [49]:
df.groupby(['id','holiday'])['visitors'].mean().diff()

id                    holiday
air_00a91d42b08b08d9  0                NaN
                      1          -5.103896
air_0164b9927d20bcc3  0         -11.708333
                      1          -1.291667
air_0241aa3964b7f861  0           1.883905
                                   ...    
air_fef9ccb3ba0da2f7  1           2.534783
air_ffcc2d5087e1b476  0           8.436975
                      1          -9.436975
air_fff68b929994bfbd  0          -5.906615
                      1          -0.093385
Name: visitors, Length: 1646, dtype: float64

In [34]:
q2 = df.groupby(['id','holiday'])['visitors'].mean().reset_index()

In [41]:
q2[(q2['holiday']==1)]

id  holiday   visitors
1     air_00a91d42b08b08d9        1  21.000000
3     air_0164b9927d20bcc3        1   8.000000
5     air_0241aa3964b7f861        1  10.176471
7     air_0328696196e46f18        1   7.166667
9     air_034a3d5b40d5b1b1        1  14.400000
...                    ...      ...        ...
1637  air_fea5dc9594450608        1  14.400000
1639  air_fee8dcf4d619598e        1  28.882353
1641  air_fef9ccb3ba0da2f7        1  12.000000
1643  air_ffcc2d5087e1b476        1  11.000000
1645  air_fff68b929994bfbd        1   5.000000

[817 rows x 3 columns]

In [42]:
q2[(q2['holiday']==0)]

id  holiday   visitors
0     air_00a91d42b08b08d9        0  26.103896
2     air_0164b9927d20bcc3        0   9.291667
4     air_0241aa3964b7f861        0   9.883905
6     air_0328696196e46f18        0   7.981818
8     air_034a3d5b40d5b1b1        0  14.855932
...                    ...      ...        ...
1636  air_fea5dc9594450608        0  14.488636
1638  air_fee8dcf4d619598e        0  25.848708
1640  air_fef9ccb3ba0da2f7        0   9.465217
1642  air_ffcc2d5087e1b476        0  20.436975
1644  air_fff68b929994bfbd        0   5.093385

[829 rows x 3 columns]

In [45]:
q2[(q2['holiday']==0)]['visitors'] - q2[(q2['holiday']==1)]['visitors']

### Suggested solution Q2:

In [76]:
restauarant_visits = df.groupby(['id','holiday'])['visitors'].mean().reset_index()

In [77]:
restauarant_visits

id  holiday   visitors
0     air_00a91d42b08b08d9        0  26.103896
1     air_00a91d42b08b08d9        1  21.000000
2     air_0164b9927d20bcc3        0   9.291667
3     air_0164b9927d20bcc3        1   8.000000
4     air_0241aa3964b7f861        0   9.883905
...                    ...      ...        ...
1641  air_fef9ccb3ba0da2f7        1  12.000000
1642  air_ffcc2d5087e1b476        0  20.436975
1643  air_ffcc2d5087e1b476        1  11.000000
1644  air_fff68b929994bfbd        0   5.093385
1645  air_fff68b929994bfbd        1   5.000000

[1646 rows x 3 columns]

In [79]:
restauarant_visits.groupby('id')['visitors'].diff().dropna()

1      -5.103896
3      -1.291667
5       0.292566
7      -0.815152
9      -0.455932
          ...   
1637   -0.088636
1639    3.033644
1641    2.534783
1643   -9.436975
1645   -0.093385
Name: visitors, Length: 817, dtype: float64

**Question 3:** Can you grab the first 15 rows of dates for each restaurant?  The last 15 rows?

In [2]:
# your answer here

In [43]:
?df.groupby

In [47]:
?df.sort_values

In [51]:
q3 = df.sort_values(by=['id','visit_date'],ascending=[True,True]).groupby('id').head(15)

In [52]:
q3.iloc[0:45,:]

id  visit_date  visitors calendar_date day_of_week  \
166836  air_00a91d42b08b08d9  2016-07-01        35    2016-07-01      Friday   
166837  air_00a91d42b08b08d9  2016-07-02         9    2016-07-02    Saturday   
166838  air_00a91d42b08b08d9  2016-07-04        20    2016-07-04      Monday   
166839  air_00a91d42b08b08d9  2016-07-05        25    2016-07-05     Tuesday   
166840  air_00a91d42b08b08d9  2016-07-06        29    2016-07-06   Wednesday   
166841  air_00a91d42b08b08d9  2016-07-07        34    2016-07-07    Thursday   
166842  air_00a91d42b08b08d9  2016-07-08        42    2016-07-08      Friday   
166843  air_00a91d42b08b08d9  2016-07-09        11    2016-07-09    Saturday   
166844  air_00a91d42b08b08d9  2016-07-11        25    2016-07-11      Monday   
166845  air_00a91d42b08b08d9  2016-07-12        24    2016-07-12     Tuesday   
166846  air_00a91d42b08b08d9  2016-07-13        36    2016-07-13   Wednesday   
166847  air_00a91d42b08b08d9  2016-07-14        28    2016-07-14    Thursday   
166848  air_00a91d42b08b08d9  2016-07-15        28    2016-07-15      Friday   
166849  air_00a91d42b08b08d9  2016-07-16         5    2016-07-16    Saturday   
166850  air_00a91d42b08b08d9  2016-07-19        25    2016-07-19     Tuesday   
250727  air_0164b9927d20bcc3  2016-10-03         3    2016-10-03      Monday   
250728  air_0164b9927d20bcc3  2016-10-04         8    2016-10-04     Tuesday   
250729  air_0164b9927d20bcc3  2016-10-05        10    2016-10-05   Wednesday   
250730  air_0164b9927d20bcc3  2016-10-06        11    2016-10-06    Thursday   
250731  air_0164b9927d20bcc3  2016-10-07        13    2016-10-07      Friday   
250732  air_0164b9927d20bcc3  2016-10-08         2    2016-10-08    Saturday   
250733  air_0164b9927d20bcc3  2016-10-11         3    2016-10-11     Tuesday   
250734  air_0164b9927d20bcc3  2016-10-12        17    2016-10-12   Wednesday   
250735  air_0164b9927d20bcc3  2016-10-13        12    2016-10-13    Thursday   
250736  air_0164b9927d20bcc3  2016-10-14        11    2016-10-14      Friday   
250737  air_0164b9927d20bcc3  2016-10-15         2    2016-10-15    Saturday   
250738  air_0164b9927d20bcc3  2016-10-18         2    2016-10-18     Tuesday   
250739  air_0164b9927d20bcc3  2016-10-19         3    2016-10-19   Wednesday   
250740  air_0164b9927d20bcc3  2016-10-20        11    2016-10-20    Thursday   
250741  air_0164b9927d20bcc3  2016-10-21        16    2016-10-21      Friday   
99657   air_0241aa3964b7f861  2016-01-03        10    2016-01-03      Sunday   
99653   air_0241aa3964b7f861  2016-01-04         9    2016-01-04      Monday   
99583   air_0241aa3964b7f861  2016-01-05        17    2016-01-05     Tuesday   
99584   air_0241aa3964b7f861  2016-01-06        10    2016-01-06   Wednesday   
99585   air_0241aa3964b7f861  2016-01-08         5    2016-01-08      Friday   
99586   air_0241aa3964b7f861  2016-01-09         8    2016-01-09    Saturday   
99587   air_0241aa3964b7f861  2016-01-10        16    2016-01-10      Sunday   
99588   air_0241aa3964b7f861  2016-01-11        16    2016-01-11      Monday   
99654   air_0241aa3964b7f861  2016-01-12         8    2016-01-12     Tuesday   
99262   air_0241aa3964b7f861  2016-01-13         9    2016-01-13   Wednesday   
99263   air_0241aa3964b7f861  2016-01-14         4    2016-01-14    Thursday   
99264   air_0241aa3964b7f861  2016-01-15         4    2016-01-15      Friday   
99265   air_0241aa3964b7f861  2016-01-16        29    2016-01-16    Saturday   
99589   air_0241aa3964b7f861  2016-01-17        10    2016-01-17      Sunday   
99266   air_0241aa3964b7f861  2016-01-18         6    2016-01-18      Monday   

        holiday           genre                             area   latitude  \
166836        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
166837        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
166838        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
166839        0  Italian/Fren

In [65]:
q3b = (df.sort_values(by=['id','visit_date'],ascending=[True,False]).groupby('id').head(15)).sort_values(by=['id','visit_date'],ascending=[True,True])

In [66]:
q3b.iloc[0:45,:]

id visit_date  visitors calendar_date day_of_week  \
167048  air_00a91d42b08b08d9 2017-04-05        35    2017-04-05   Wednesday   
167049  air_00a91d42b08b08d9 2017-04-06        29    2017-04-06    Thursday   
167050  air_00a91d42b08b08d9 2017-04-07        17    2017-04-07      Friday   
167051  air_00a91d42b08b08d9 2017-04-08         9    2017-04-08    Saturday   
167052  air_00a91d42b08b08d9 2017-04-10        17    2017-04-10      Monday   
167053  air_00a91d42b08b08d9 2017-04-11        43    2017-04-11     Tuesday   
167054  air_00a91d42b08b08d9 2017-04-12        28    2017-04-12   Wednesday   
167055  air_00a91d42b08b08d9 2017-04-13        34    2017-04-13    Thursday   
167056  air_00a91d42b08b08d9 2017-04-14        39    2017-04-14      Friday   
167057  air_00a91d42b08b08d9 2017-04-17        19    2017-04-17      Monday   
167058  air_00a91d42b08b08d9 2017-04-18        35    2017-04-18     Tuesday   
167059  air_00a91d42b08b08d9 2017-04-19        17    2017-04-19   Wednesday   
167060  air_00a91d42b08b08d9 2017-04-20        38    2017-04-20    Thursday   
167061  air_00a91d42b08b08d9 2017-04-21        55    2017-04-21      Friday   
167062  air_00a91d42b08b08d9 2017-04-22        18    2017-04-22    Saturday   
250860  air_0164b9927d20bcc3 2017-04-06         6    2017-04-06    Thursday   
250861  air_0164b9927d20bcc3 2017-04-07         6    2017-04-07      Friday   
250862  air_0164b9927d20bcc3 2017-04-08         1    2017-04-08    Saturday   
250863  air_0164b9927d20bcc3 2017-04-10        10    2017-04-10      Monday   
250864  air_0164b9927d20bcc3 2017-04-11         8    2017-04-11     Tuesday   
250865  air_0164b9927d20bcc3 2017-04-12         2    2017-04-12   Wednesday   
250866  air_0164b9927d20bcc3 2017-04-13        13    2017-04-13    Thursday   
250867  air_0164b9927d20bcc3 2017-04-14         7    2017-04-14      Friday   
250868  air_0164b9927d20bcc3 2017-04-15         1    2017-04-15    Saturday   
250869  air_0164b9927d20bcc3 2017-04-17         2    2017-04-17      Monday   
250870  air_0164b9927d20bcc3 2017-04-18         1    2017-04-18     Tuesday   
250871  air_0164b9927d20bcc3 2017-04-19         8    2017-04-19   Wednesday   
250872  air_0164b9927d20bcc3 2017-04-20         1    2017-04-20    Thursday   
250873  air_0164b9927d20bcc3 2017-04-21        26    2017-04-21      Friday   
250874  air_0164b9927d20bcc3 2017-04-22         6    2017-04-22    Saturday   
99569   air_0241aa3964b7f861 2017-04-05         7    2017-04-05   Wednesday   
99570   air_0241aa3964b7f861 2017-04-07        12    2017-04-07      Friday   
99571   air_0241aa3964b7f861 2017-04-08         4    2017-04-08    Saturday   
99647   air_0241aa3964b7f861 2017-04-09        10    2017-04-09      Sunday   
99572   air_0241aa3964b7f861 2017-04-10        11    2017-04-10      Monday   
99573   air_0241aa3964b7f861 2017-04-11         5    2017-04-11     Tuesday   
99574   air_0241aa3964b7f861 2017-04-12        13    2017-04-12   Wednesday   
99575   air_0241aa3964b7f861 2017-04-14         4    2017-04-14      Friday   
99576   air_0241aa3964b7f861 2017-04-15        15    2017-04-15    Saturday   
99577   air_0241aa3964b7f861 2017-04-16        10    2017-04-16      Sunday   
99578   air_0241aa3964b7f861 2017-04-17        12    2017-04-17      Monday   
99579   air_0241aa3964b7f861 2017-04-18        19    2017-04-18     Tuesday   
99580   air_0241aa3964b7f861 2017-04-19         8    2017-04-19   Wednesday   
99581   air_0241aa3964b7f861 2017-04-21         3    2017-04-21      Friday   
99582   air_0241aa3964b7f861 2017-04-22        13    2017-04-22    Saturday   

        holiday           genre                             area   latitude  \
167048        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
167049        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
167050        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
167051        0  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami  35.694003

`(
    df.sort_values(
       by=['id','visit_date'],
       ascending=True,False]
       )
       .groupby('id')
       .head(15)
)
.sort_values(by=['id','visit_date'],ascending=[True,True])`

So

*1:* Sort the values so that visit date is ordered descending

*2:* Group all the values with the same ID together in the data set

*3:* Take the first 15 rows with head

*4:* Convert to it's own data frame

*5:* Resort the resulting groups of 15 rows so that the dates are in ascending order

### Jonathan's approach (Q3):

In [6]:
df.iloc[:15]

id  visit_date  visitors calendar_date day_of_week  \
0   air_ba937bf13d40fb24  2016-01-13        25    2016-01-13   Wednesday   
1   air_ba937bf13d40fb24  2016-01-14        32    2016-01-14    Thursday   
2   air_ba937bf13d40fb24  2016-01-15        29    2016-01-15      Friday   
3   air_ba937bf13d40fb24  2016-01-16        22    2016-01-16    Saturday   
4   air_ba937bf13d40fb24  2016-01-18         6    2016-01-18      Monday   
5   air_ba937bf13d40fb24  2016-01-19         9    2016-01-19     Tuesday   
6   air_ba937bf13d40fb24  2016-01-20        31    2016-01-20   Wednesday   
7   air_ba937bf13d40fb24  2016-01-21        21    2016-01-21    Thursday   
8   air_ba937bf13d40fb24  2016-01-22        18    2016-01-22      Friday   
9   air_ba937bf13d40fb24  2016-01-23        26    2016-01-23    Saturday   
10  air_ba937bf13d40fb24  2016-01-25        21    2016-01-25      Monday   
11  air_ba937bf13d40fb24  2016-01-26        11    2016-01-26     Tuesday   
12  air_ba937bf13d40fb24  2016-01-27        24    2016-01-27   Wednesday   
13  air_ba937bf13d40fb24  2016-01-28        21    2016-01-28    Thursday   
14  air_ba937bf13d40fb24  2016-01-29        26    2016-01-29      Friday   

    holiday       genre                          area   latitude   longitude  \
0         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
5         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
6         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
7         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
8         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
9         0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
10        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
11        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
12        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
13        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
14        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

    reserve_visitors  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  
5                NaN  
6                NaN  
7                NaN  
8                NaN  
9                NaN  
10               NaN  
11               2.0  
12               NaN  
13               NaN  
14               NaN

In [15]:
restaurant_vals = df.groupby('id').apply(lambda x: x.iloc[:15])

# This is effectively creating an additional index based on the id on the right

In [14]:
restaurant_vals[:17]

index                    id visit_date  visitors  \
id                                                                           
air_00a91d42b08b08d9 0   166836  air_00a91d42b08b08d9 2016-07-01        35   
                     1   166837  air_00a91d42b08b08d9 2016-07-02         9   
                     2   166838  air_00a91d42b08b08d9 2016-07-04        20   
                     3   166839  air_00a91d42b08b08d9 2016-07-05        25   
                     4   166840  air_00a91d42b08b08d9 2016-07-06        29   
                     5   166841  air_00a91d42b08b08d9 2016-07-07        34   
                     6   166842  air_00a91d42b08b08d9 2016-07-08        42   
                     7   166843  air_00a91d42b08b08d9 2016-07-09        11   
                     8   166844  air_00a91d42b08b08d9 2016-07-11        25   
                     9   166845  air_00a91d42b08b08d9 2016-07-12        24   
                     10  166846  air_00a91d42b08b08d9 2016-07-13        36   
                     11  166847  air_00a91d42b08b08d9 2016-07-14        28   
                     12  166848  air_00a91d42b08b08d9 2016-07-15        28   
                     13  166849  air_00a91d42b08b08d9 2016-07-16         5   
                     14  166850  air_00a91d42b08b08d9 2016-07-19        25   
air_0164b9927d20bcc3 0   250727  air_0164b9927d20bcc3 2016-10-03         3   
                     1   250728  air_0164b9927d20bcc3 2016-10-04         8   

                        calendar_date day_of_week  holiday           genre  \
id                                                                           
air_00a91d42b08b08d9 0     2016-07-01      Friday        0  Italian/French   
                     1     2016-07-02    Saturday        0  Italian/French   
                     2     2016-07-04      Monday        0  Italian/French   
                     3     2016-07-05     Tuesday        0  Italian/French   
                     4     2016-07-06   Wednesday        0  Italian/French   
                     5     2016-07-07    Thursday        0  Italian/French   
                     6     2016-07-08      Friday        0  Italian/French   
                     7     2016-07-09    Saturday        0  Italian/French   
                     8     2016-07-11      Monday        0  Italian/French   
                     9     2016-07-12     Tuesday        0  Italian/French   
                     10    2016-07-13   Wednesday        0  Italian/French   
                     11    2016-07-14    Thursday        0  Italian/French   
                     12    2016-07-15      Friday        0  Italian/French   
                     13    2016-07-16    Saturday        0  Italian/French   
                     14    2016-07-19     Tuesday        0  Italian/French   
air_0164b9927d20bcc3 0     2016-10-03      Monday        0  Italian/French   
                     1     2016-10-04     Tuesday        0  Italian/French   

                                                    area   latitude  \
id                                                                    
air_00a91d42b08b08d9 0   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     1   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     2   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     3   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     4   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     5   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     6   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     7   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     8   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     9   Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     10  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     11  Tōkyō-to Chiyoda-ku Kudanminami  35.694003   
                     12  Tōkyō-to Chiyoda-ku Kudanminami  35.69400

In [19]:
restaurant_vals_rev = df.groupby('id').apply(lambda x: x.iloc[-15:])

In [20]:
restaurant_vals_rev[:20]

id visit_date  visitors  \
id                                                                       
air_00a91d42b08b08d9 167053  air_00a91d42b08b08d9 2017-04-11        43   
                     167054  air_00a91d42b08b08d9 2017-04-12        28   
                     167055  air_00a91d42b08b08d9 2017-04-13        34   
                     167056  air_00a91d42b08b08d9 2017-04-14        39   
                     167057  air_00a91d42b08b08d9 2017-04-17        19   
                     167058  air_00a91d42b08b08d9 2017-04-18        35   
                     167059  air_00a91d42b08b08d9 2017-04-19        17   
                     167060  air_00a91d42b08b08d9 2017-04-20        38   
                     167061  air_00a91d42b08b08d9 2017-04-21        55   
                     167062  air_00a91d42b08b08d9 2017-04-22        18   
                     167063  air_00a91d42b08b08d9 2016-08-17        30   
                     167064  air_00a91d42b08b08d9 2016-08-18        20   
                     167065  air_00a91d42b08b08d9 2016-08-15        21   
                     167066  air_00a91d42b08b08d9 2016-08-16        12   
                     167067  air_00a91d42b08b08d9 2017-01-04        17   
air_0164b9927d20bcc3 250861  air_0164b9927d20bcc3 2017-04-07         6   
                     250862  air_0164b9927d20bcc3 2017-04-08         1   
                     250863  air_0164b9927d20bcc3 2017-04-10        10   
                     250864  air_0164b9927d20bcc3 2017-04-11         8   
                     250865  air_0164b9927d20bcc3 2017-04-12         2   

                            calendar_date day_of_week  holiday  \
id                                                               
air_00a91d42b08b08d9 167053    2017-04-11     Tuesday        0   
                     167054    2017-04-12   Wednesday        0   
                     167055    2017-04-13    Thursday        0   
                     167056    2017-04-14      Friday        0   
                     167057    2017-04-17      Monday        0   
                     167058    2017-04-18     Tuesday        0   
                     167059    2017-04-19   Wednesday        0   
                     167060    2017-04-20    Thursday        0   
                     167061    2017-04-21      Friday        0   
                     167062    2017-04-22    Saturday        0   
                     167063    2016-08-17   Wednesday        0   
                     167064    2016-08-18    Thursday        0   
                     167065    2016-08-15      Monday        1   
                     167066    2016-08-16     Tuesday        0   
                     167067    2017-01-04   Wednesday        0   
air_0164b9927d20bcc3 250861    2017-04-07      Friday        0   
                     250862    2017-04-08    Saturday        0   
                     250863    2017-04-10      Monday        0   
                     250864    2017-04-11     Tuesday        0   
                     250865    2017-04-12   Wednesday        0   

                                      genre                             area  \
id                                                                             
air_00a91d42b08b08d9 167053  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167054  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167055  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167056  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167057  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167058  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167059  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167060  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167061  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
                     167062  Italian/French  Tōkyō-to Chiyoda-ku Kudanminami   
     

**Question 4:** Grab the quarterley sales for each individual restaurant within our dataset

In [3]:
# your answer here

In [69]:
df.head()

id visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN

In [138]:
df['visitors'].isnull().sum()

0

In [137]:
df['reserve_visitors'].isnull().sum()

143714

In [ ]:
# interpreting quarterly sales, as number of visitors each quarter

In [70]:
df.visit_date.dt.year

0         2016
1         2016
2         2016
3         2016
4         2016
          ... 
252103    2017
252104    2017
252105    2017
252106    2017
252107    2017
Name: visit_date, Length: 252108, dtype: int64

In [68]:
df.visit_date.dt.quarter

0         1
1         1
2         1
3         1
4         1
         ..
252103    2
252104    2
252105    1
252106    1
252107    2
Name: visit_date, Length: 252108, dtype: int64

In [126]:
q4 = df.groupby(by=['id',df.visit_date.dt.year,df.visit_date.dt.quarter])['visitors'].sum()

`df.groupby(by=['id',df.visit_date.dt.year,df.visit_date.dt.quarter])['visitors'].sum()`
The above works in terms of generating an output.

The problem is that it runs into an error when resetting the index.
So need to rename the columns (somehow) within the series.


The way to do that was in the end simple: `q4df.index.names = ['id','visit_year','visit_date']`. Then because the names were changed we could use `reset_index()` to put them all at the same level.


In [140]:
df \
.groupby(by=['id',df.visit_date.dt.year,df.visit_date.dt.quarter]).sum()['visitors']

id                    visit_date  visit_date
air_00a91d42b08b08d9  2016        3             1780
                                  4             1740
                      2017        1             2041
                                  2              490
air_0164b9927d20bcc3  2016        4              627
                                                ... 
air_ffcc2d5087e1b476  2017        2              390
air_fff68b929994bfbd  2016        3              404
                                  4              452
                      2017        1              411
                                  2              102
Name: visitors, Length: 3903, dtype: int64

In [110]:
q4

id                    visit_date  visit_date
air_00a91d42b08b08d9  2016        3             1780
                                  4             1740
                      2017        1             2041
                                  2              490
air_0164b9927d20bcc3  2016        4              627
                                                ... 
air_ffcc2d5087e1b476  2017        2              390
air_fff68b929994bfbd  2016        3              404
                                  4              452
                      2017        1              411
                                  2              102
Name: visitors, Length: 3903, dtype: int64

In [111]:
type(q4)

pandas.core.series.Series

In [94]:
?pd.DataFrame

In [115]:
q4df = q4.to_frame('quarterly_visits')

In [157]:
q4df

quarterly_visits
id                   visit_date visit_date                  
air_00a91d42b08b08d9 2016       3                       1780
                                4                       1740
                     2017       1                       2041
                                2                        490
air_0164b9927d20bcc3 2016       4                        627
...                                                      ...
air_ffcc2d5087e1b476 2017       2                        390
air_fff68b929994bfbd 2016       3                        404
                                4                        452
                     2017       1                        411
                                2                        102

[3903 rows x 1 columns]

In [166]:
q4df.index

MultiIndex([('air_00a91d42b08b08d9', 2016, 3),
            ('air_00a91d42b08b08d9', 2016, 4),
            ('air_00a91d42b08b08d9', 2017, 1),
            ('air_00a91d42b08b08d9', 2017, 2),
            ('air_0164b9927d20bcc3', 2016, 4),
            ('air_0164b9927d20bcc3', 2017, 1),
            ('air_0164b9927d20bcc3', 2017, 2),
            ('air_0241aa3964b7f861', 2016, 1),
            ('air_0241aa3964b7f861', 2016, 2),
            ('air_0241aa3964b7f861', 2016, 3),
            ...
            ('air_fef9ccb3ba0da2f7', 2017, 1),
            ('air_fef9ccb3ba0da2f7', 2017, 2),
            ('air_ffcc2d5087e1b476', 2016, 3),
            ('air_ffcc2d5087e1b476', 2016, 4),
            ('air_ffcc2d5087e1b476', 2017, 1),
            ('air_ffcc2d5087e1b476', 2017, 2),
            ('air_fff68b929994bfbd', 2016, 3),
            ('air_fff68b929994bfbd', 2016, 4),
            ('air_fff68b929994bfbd', 2017, 1),
            ('air_fff68b929994bfbd', 2017, 2)],
           names=['id', 'visit_date', 'visi

In [172]:
q4df.index.names = ['id','visit_year','visit_date']

In [173]:
q4df.index

MultiIndex([('air_00a91d42b08b08d9', 2016, 3),
            ('air_00a91d42b08b08d9', 2016, 4),
            ('air_00a91d42b08b08d9', 2017, 1),
            ('air_00a91d42b08b08d9', 2017, 2),
            ('air_0164b9927d20bcc3', 2016, 4),
            ('air_0164b9927d20bcc3', 2017, 1),
            ('air_0164b9927d20bcc3', 2017, 2),
            ('air_0241aa3964b7f861', 2016, 1),
            ('air_0241aa3964b7f861', 2016, 2),
            ('air_0241aa3964b7f861', 2016, 3),
            ...
            ('air_fef9ccb3ba0da2f7', 2017, 1),
            ('air_fef9ccb3ba0da2f7', 2017, 2),
            ('air_ffcc2d5087e1b476', 2016, 3),
            ('air_ffcc2d5087e1b476', 2016, 4),
            ('air_ffcc2d5087e1b476', 2017, 1),
            ('air_ffcc2d5087e1b476', 2017, 2),
            ('air_fff68b929994bfbd', 2016, 3),
            ('air_fff68b929994bfbd', 2016, 4),
            ('air_fff68b929994bfbd', 2017, 1),
            ('air_fff68b929994bfbd', 2017, 2)],
           names=['id', 'visit_year', 'visi

In [175]:
q4df_final = q4df.reset_index()

In [176]:
q4df_final

id  visit_year  visit_date  quarterly_visits
0     air_00a91d42b08b08d9        2016           3              1780
1     air_00a91d42b08b08d9        2016           4              1740
2     air_00a91d42b08b08d9        2017           1              2041
3     air_00a91d42b08b08d9        2017           2               490
4     air_0164b9927d20bcc3        2016           4               627
...                    ...         ...         ...               ...
3898  air_ffcc2d5087e1b476        2017           2               390
3899  air_fff68b929994bfbd        2016           3               404
3900  air_fff68b929994bfbd        2016           4               452
3901  air_fff68b929994bfbd        2017           1               411
3902  air_fff68b929994bfbd        2017           2               102

[3903 rows x 4 columns]

### Jonathan's approach (Q4):

In [21]:
df.groupby(['id',df.visit_date.dt.year, df.visit_date.dt.quarter])['visitors'].sum()

id                    visit_date  visit_date
air_00a91d42b08b08d9  2016        3             1780
                                  4             1740
                      2017        1             2041
                                  2              490
air_0164b9927d20bcc3  2016        4              627
                                                ... 
air_ffcc2d5087e1b476  2017        2              390
air_fff68b929994bfbd  2016        3              404
                                  4              452
                      2017        1              411
                                  2              102
Name: visitors, Length: 3903, dtype: int64

In [22]:
df.groupby(['id',df.visit_date.dt.year, df.visit_date.dt.quarter])['visitors'].sum().index

MultiIndex([('air_00a91d42b08b08d9', 2016, 3),
            ('air_00a91d42b08b08d9', 2016, 4),
            ('air_00a91d42b08b08d9', 2017, 1),
            ('air_00a91d42b08b08d9', 2017, 2),
            ('air_0164b9927d20bcc3', 2016, 4),
            ('air_0164b9927d20bcc3', 2017, 1),
            ('air_0164b9927d20bcc3', 2017, 2),
            ('air_0241aa3964b7f861', 2016, 1),
            ('air_0241aa3964b7f861', 2016, 2),
            ('air_0241aa3964b7f861', 2016, 3),
            ...
            ('air_fef9ccb3ba0da2f7', 2017, 1),
            ('air_fef9ccb3ba0da2f7', 2017, 2),
            ('air_ffcc2d5087e1b476', 2016, 3),
            ('air_ffcc2d5087e1b476', 2016, 4),
            ('air_ffcc2d5087e1b476', 2017, 1),
            ('air_ffcc2d5087e1b476', 2017, 2),
            ('air_fff68b929994bfbd', 2016, 3),
            ('air_fff68b929994bfbd', 2016, 4),
            ('air_fff68b929994bfbd', 2017, 1),
            ('air_fff68b929994bfbd', 2017, 2)],
           names=['id', 'visit_date', 'visi

In [24]:
type(df.groupby(['id',df.visit_date.dt.year, df.visit_date.dt.quarter])['visitors'].sum())

# A series is what comes back - it's a series with an index 3 levels deep.
# The groupby creates index levels rather than columns. It helps to organise the data

pandas.core.series.Series

In [ ]:
# Getting it into a data frame

In [25]:
df['year'] = df['visit_date'].dt.year
df['quarter'] = df['visit_date'].dt.quarter

In [26]:
df.groupby(['id','year','quarter'])['visitors'].sum().reset_index()

id  year  quarter  visitors
0     air_00a91d42b08b08d9  2016        3      1780
1     air_00a91d42b08b08d9  2016        4      1740
2     air_00a91d42b08b08d9  2017        1      2041
3     air_00a91d42b08b08d9  2017        2       490
4     air_0164b9927d20bcc3  2016        4       627
...                    ...   ...      ...       ...
3898  air_ffcc2d5087e1b476  2017        2       390
3899  air_fff68b929994bfbd  2016        3       404
3900  air_fff68b929994bfbd  2016        4       452
3901  air_fff68b929994bfbd  2017        1       411
3902  air_fff68b929994bfbd  2017        2       102

[3903 rows x 4 columns]

**Question 6:** What restaurant had the highest amount of reservations?

In [4]:
# your answer here

In [183]:
df.groupby('id').sum().sort_values(by='visitors',ascending=False)['visitors']


id
air_399904bdb7685ca0    18717
air_f26f36ec4dc5adb0    18577
air_e55abd740f93ecc4    18101
air_99157b6163835eec    18097
air_5c817ef28f236bdf    18009
                        ...  
air_9dd7d38b0f1760c4      803
air_5b704df317ed1962      800
air_fdcfef8bd859f650      625
air_bbe1c1a47e09f161      581
air_a21ffca0bea1661a      190
Name: visitors, Length: 829, dtype: int64

In [ ]:
air_399904bdb7685ca0

In [187]:
df.groupby('id').sum().idxmax()['visitors']

'air_399904bdb7685ca0'

In [ ]:
# Jonathan's answers

In [29]:
df.groupby('id')['visitors'].sum().idxmax

<bound method Series.idxmax of id
air_00a91d42b08b08d9    6051
air_0164b9927d20bcc3    1378
air_0241aa3964b7f861    3919
air_0328696196e46f18     921
air_034a3d5b40d5b1b1    3722
                        ... 
air_fea5dc9594450608    3969
air_fee8dcf4d619598e    7496
air_fef9ccb3ba0da2f7    2357
air_ffcc2d5087e1b476    4919
air_fff68b929994bfbd    1369
Name: visitors, Length: 829, dtype: int64>

**Question 7:** What is the total number of missing entries for each restaurant?  

In [5]:
# your answer here

In [191]:
(df['visit_date'].max()-df['visit_date'].min()).days

477

So 477 is the maximum number of entries per restaurant

In [199]:
df.groupby('id').count()['visit_date']

id
air_00a91d42b08b08d9    232
air_0164b9927d20bcc3    149
air_0241aa3964b7f861    396
air_0328696196e46f18    116
air_034a3d5b40d5b1b1    251
                       ... 
air_fea5dc9594450608    274
air_fee8dcf4d619598e    288
air_fef9ccb3ba0da2f7    245
air_ffcc2d5087e1b476    243
air_fff68b929994bfbd    269
Name: visit_date, Length: 829, dtype: int64

In [202]:
q7 = df.groupby('id').apply(lambda x: (df['visit_date'].max()-df['visit_date'].min()).days - x.count()['visit_date'])

In [203]:
q7.reset_index()

id    0
0    air_00a91d42b08b08d9  245
1    air_0164b9927d20bcc3  328
2    air_0241aa3964b7f861   81
3    air_0328696196e46f18  361
4    air_034a3d5b40d5b1b1  226
..                    ...  ...
824  air_fea5dc9594450608  203
825  air_fee8dcf4d619598e  189
826  air_fef9ccb3ba0da2f7  232
827  air_ffcc2d5087e1b476  234
828  air_fff68b929994bfbd  208

[829 rows x 2 columns]

In [ ]:
# Jonathan's answer

In [30]:
df.isnull().sum()

id                       0
visit_date               0
visitors                 0
calendar_date            0
day_of_week              0
holiday                  0
genre                    0
area                     0
latitude                 0
longitude                0
reserve_visitors    143714
year                     0
quarter                  0
dtype: int64

In [31]:
# Can't do this
df.groupby('id').isnull().sum()

AttributeError: 'DataFrameGroupBy' object has no attribute 'isnull'

In [35]:
df.groupby('id').apply(lambda x: x.isnull().sum())

id  visit_date  visitors  calendar_date  day_of_week  \
id                                                                           
air_00a91d42b08b08d9   0           0         0              0            0   
air_0164b9927d20bcc3   0           0         0              0            0   
air_0241aa3964b7f861   0           0         0              0            0   
air_0328696196e46f18   0           0         0              0            0   
air_034a3d5b40d5b1b1   0           0         0              0            0   
...                   ..         ...       ...            ...          ...   
air_fea5dc9594450608   0           0         0              0            0   
air_fee8dcf4d619598e   0           0         0              0            0   
air_fef9ccb3ba0da2f7   0           0         0              0            0   
air_ffcc2d5087e1b476   0           0         0              0            0   
air_fff68b929994bfbd   0           0         0              0            0   

                      holiday  genre  area  latitude  longitude  \
id                                                                
air_00a91d42b08b08d9        0      0     0         0          0   
air_0164b9927d20bcc3        0      0     0         0          0   
air_0241aa3964b7f861        0      0     0         0          0   
air_0328696196e46f18        0      0     0         0          0   
air_034a3d5b40d5b1b1        0      0     0         0          0   
...                       ...    ...   ...       ...        ...   
air_fea5dc9594450608        0      0     0         0          0   
air_fee8dcf4d619598e        0      0     0         0          0   
air_fef9ccb3ba0da2f7        0      0     0         0          0   
air_ffcc2d5087e1b476        0      0     0         0          0   
air_fff68b929994bfbd        0      0     0         0          0   

                      reserve_visitors  year  quarter  
id                                                     
air_00a91d42b08b08d9               122     0        0  
air_0164b9927d20bcc3                50     0        0  
air_0241aa3964b7f861               249     0        0  
air_0328696196e46f18                50     0        0  
air_034a3d5b40d5b1b1               130     0        0  
...                                ...   ...      ...  
air_fea5dc9594450608               139     0        0  
air_fee8dcf4d619598e               149     0        0  
air_fef9ccb3ba0da2f7               123     0        0  
air_ffcc2d5087e1b476               124     0        0  
air_fff68b929994bfbd               132     0        0  

[829 rows x 13 columns]

In [36]:

df.groupby('id').apply(lambda x: x.isnull().sum())['reserve_visitors']

id
air_00a91d42b08b08d9    122
air_0164b9927d20bcc3     50
air_0241aa3964b7f861    249
air_0328696196e46f18     50
air_034a3d5b40d5b1b1    130
                       ... 
air_fea5dc9594450608    139
air_fee8dcf4d619598e    149
air_fef9ccb3ba0da2f7    123
air_ffcc2d5087e1b476    124
air_fff68b929994bfbd    132
Name: reserve_visitors, Length: 829, dtype: int64

In [38]:
# This adds it up for each of the individual grouped data sets.
df.groupby('id').apply(lambda x: x.isnull().sum().sum()) 

id
air_00a91d42b08b08d9    122
air_0164b9927d20bcc3     50
air_0241aa3964b7f861    249
air_0328696196e46f18     50
air_034a3d5b40d5b1b1    130
                       ... 
air_fea5dc9594450608    139
air_fee8dcf4d619598e    149
air_fef9ccb3ba0da2f7    123
air_ffcc2d5087e1b476    124
air_fff68b929994bfbd    132
Length: 829, dtype: int64

**Question 8:**  Create two variables, `train` and `test`.  Make `train` a dataset that contains all but the **last 15 rows** for each restaurant.  Make `test` the last 15 rows for each restaurant.

In [6]:
# your answer here

In [215]:
(df.assign(rn=df.sort_values(['visit_date']).groupby('id').cumcount()+1)).sort_values(by=['id','rn']).iloc[:500]

id visit_date  visitors calendar_date day_of_week  \
166836  air_00a91d42b08b08d9 2016-07-01        35    2016-07-01      Friday   
166837  air_00a91d42b08b08d9 2016-07-02         9    2016-07-02    Saturday   
166838  air_00a91d42b08b08d9 2016-07-04        20    2016-07-04      Monday   
166839  air_00a91d42b08b08d9 2016-07-05        25    2016-07-05     Tuesday   
166840  air_00a91d42b08b08d9 2016-07-06        29    2016-07-06   Wednesday   
166841  air_00a91d42b08b08d9 2016-07-07        34    2016-07-07    Thursday   
166842  air_00a91d42b08b08d9 2016-07-08        42    2016-07-08      Friday   
166843  air_00a91d42b08b08d9 2016-07-09        11    2016-07-09    Saturday   
166844  air_00a91d42b08b08d9 2016-07-11        25    2016-07-11      Monday   
166845  air_00a91d42b08b08d9 2016-07-12        24    2016-07-12     Tuesday   
166846  air_00a91d42b08b08d9 2016-07-13        36    2016-07-13   Wednesday   
166847  air_00a91d42b08b08d9 2016-07-14        28    2016-07-14    Thursday   
166848  air_00a91d42b08b08d9 2016-07-15        28    2016-07-15      Friday   
166849  air_00a91d42b08b08d9 2016-07-16         5    2016-07-16    Saturday   
166850  air_00a91d42b08b08d9 2016-07-19        25    2016-07-19     Tuesday   
166851  air_00a91d42b08b08d9 2016-07-20        19    2016-07-20   Wednesday   
166852  air_00a91d42b08b08d9 2016-07-21        47    2016-07-21    Thursday   
166853  air_00a91d42b08b08d9 2016-07-22        53    2016-07-22      Friday   
166854  air_00a91d42b08b08d9 2016-07-23        16    2016-07-23    Saturday   
166855  air_00a91d42b08b08d9 2016-07-25        16    2016-07-25      Monday   
166856  air_00a91d42b08b08d9 2016-07-26        16    2016-07-26     Tuesday   
166857  air_00a91d42b08b08d9 2016-07-27        24    2016-07-27   Wednesday   
166858  air_00a91d42b08b08d9 2016-07-28        34    2016-07-28    Thursday   
166859  air_00a91d42b08b08d9 2016-07-29        31    2016-07-29      Friday   
166860  air_00a91d42b08b08d9 2016-07-30        20    2016-07-30    Saturday   
166861  air_00a91d42b08b08d9 2016-08-01        16    2016-08-01      Monday   
166862  air_00a91d42b08b08d9 2016-08-02        24    2016-08-02     Tuesday   
166863  air_00a91d42b08b08d9 2016-08-03        30    2016-08-03   Wednesday   
166864  air_00a91d42b08b08d9 2016-08-04        30    2016-08-04    Thursday   
166865  air_00a91d42b08b08d9 2016-08-05        42    2016-08-05      Friday   
166866  air_00a91d42b08b08d9 2016-08-06         7    2016-08-06    Saturday   
166867  air_00a91d42b08b08d9 2016-08-08        27    2016-08-08      Monday   
166868  air_00a91d42b08b08d9 2016-08-09        27    2016-08-09     Tuesday   
166869  air_00a91d42b08b08d9 2016-08-10        33    2016-08-10   Wednesday   
167065  air_00a91d42b08b08d9 2016-08-15        21    2016-08-15      Monday   
167066  air_00a91d42b08b08d9 2016-08-16        12    2016-08-16     Tuesday   
167063  air_00a91d42b08b08d9 2016-08-17        30    2016-08-17   Wednesday   
167064  air_00a91d42b08b08d9 2016-08-18        20    2016-08-18    Thursday   
166870  air_00a91d42b08b08d9 2016-08-19        43    2016-08-19      Friday   
166871  air_00a91d42b08b08d9 2016-08-20         6    2016-08-20    Saturday   
166872  air_00a91d42b08b08d9 2016-08-22        13    2016-08-22      Monday   
166873  air_00a91d42b08b08d9 2016-08-23        23    2016-08-23     Tuesday   
166874  air_00a91d42b08b08d9 2016-08-24        21    2016-08-24   Wednesday   
166875  air_00a91d42b08b08d9 2016-08-25        34    2016-08-25    Thursday   
166876  air_00a91d42b08b08d9 2016-08-26        39    2016-08-26      Friday   
166877  air_00a91d42b08b08d9 2016-08-27         9    2016-08-27    Saturday   
166878  air_00a91d42b08b08d9 2016-08-29        20    2016-08-29      Monday   
166879  air_00a91d42b08b08d9 2016-08-30        17    2016-08-30     Tuesday   
166880  air_00a91d42b08b08d9 2016-08-31        36    2016-08-31   Wednesday   
166881  air_00a91d42b08b08d9 2016-09-01        24    2016-09-01    Thursday 

In [216]:
q8 = (df.assign(rn=df.sort_values(['visit_date']).groupby('id').cumcount()+1)).sort_values(by=['id','rn'])

In [222]:
q8.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 252108 entries, 166836 to 216647
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                252108 non-null  object        
 1   visit_date        252108 non-null  datetime64[ns]
 2   visitors          252108 non-null  int64         
 3   calendar_date     252108 non-null  object        
 4   day_of_week       252108 non-null  object        
 5   holiday           252108 non-null  int64         
 6   genre             252108 non-null  object        
 7   area              252108 non-null  object        
 8   latitude          252108 non-null  float64       
 9   longitude         252108 non-null  float64       
 10  reserve_visitors  108394 non-null  float64       
 11  rn                252108 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(3), object(5)
memory usage: 35.0+ MB


In [234]:
q8b = (q8.groupby('id').max()['rn']).to_frame('max_rows').reset_index()

In [235]:
q8b

id  max_rows
0    air_00a91d42b08b08d9       232
1    air_0164b9927d20bcc3       149
2    air_0241aa3964b7f861       396
3    air_0328696196e46f18       116
4    air_034a3d5b40d5b1b1       251
5    air_036d4f1ee7285390       281
6    air_0382c794b73b51ad       298
7    air_03963426c9312048       429
8    air_04341b588bde96cd       472
9    air_049f6d5b402a31b2       258
10   air_04cae7c1bc9b2a0b       291
11   air_0585011fa179bcce       273
12   air_05c325d315cc17f5       468
13   air_0647f17b4dc041c8       287
14   air_064e203265ee5753       222
15   air_066f0221b8a4d533       277
16   air_06f95ac5c33aca10       401
17   air_0728814bd98f7367       431
18   air_0768ab3910f7967f       329
19   air_07b314d83059c4d2       294
20   air_07bb665f9cdfbdfb       380
21   air_082908692355165e       428
22   air_083ddc520ea47e1e       205
23   air_0845d8395f30c6bb       281
24   air_084d98859256acf0       240
25   air_0867f7bebad6a649       220
26   air_08ba8cd01b3ba010       459
27   air_08cb3c4ee6cd6a22       289
28   air_08ef81d5b7a0d13f       240
29   air_08f994758a1e76d4       289
30   air_09040f6df960ddb8       227
31   air_0919d54f0c9a24b8       234
32   air_09661c0f3259cc04       286
33   air_09a845d5b5944b01       407
34   air_09fd1f5c58583141       233
35   air_0a74a5408a0b8642       263
36   air_0b184ec04c741a6a       381
37   air_0b1e72d2d4422b20       258
38   air_0b9038300f8b2b50       292
39   air_0e1eae99b8723bc1       367
40   air_0e7c11b9abc50163       219
41   air_0ead98dd07e7a82a       138
42   air_0f0cdeee6c9bf3d7       267
43   air_0f2f96335f274801       385
44   air_0f60e1576a7d397d       195
45   air_1033310359ceeac1       356
46   air_10393f12e9069760       289
47   air_105a7954e32dba9b       192
48   air_10713fbf3071c361       269
49   air_10bbe8acd943d8f6       292
50   air_12c4fb7a423df20d       292
51   air_138ee734ac79ff90       397
52   air_138ff410757b845f       253
53   air_1408dd53f31a8a65       296
54   air_142e78ba7001da9c       430
55   air_1509881b22965b34       193
56   air_152c1f08d7d20e07       166
57   air_15ae33469e9ea2dd       232
58   air_15e6e15c7ea2c162       280
59   air_16179d43b6ee5fd8       232
60   air_1653a6c513865af3       424
61   air_168441ada3e878e1       293
62   air_16c4cfddeb2cf69b       231
63   air_16cf0a73233896de       410
64   air_1707a3f18bb0da07       245
65   air_17a6ab40f97fd4d8       115
66   air_17bed6dbf7c8b0fc       315
67   air_1979eaff8189d086       246
68   air_1ab60ce33bfed8a8       263
69   air_1ae94f514a0bce13       227
70   air_1ba4e87ef7422183       286
71   air_1c0b150f9e696a5f        51
72   air_1c95a84924d72500       460
73   air_1d1e8860ae04f8e9       425
74   air_1d25ca6c76df48b4       287
75   air_1d3f797dd1f7cf1c       287
76   air_1dd8f6f47480d1a2       267
77   air_1dea9815ccd36620       246
78   air_1e23210b584540e7       451
79   air_1e665503b8474c55       468
80   air_1eeff462acb24fb7       258
81   air_1f1390a8be2272b3       425
82   air_1f34e9beded2231a       148
83   air_1f7f8fa557bc0d55       424
84   air_2009041dbf9264de       409
85   air_20619d21192aa571       426
86   air_20add8092c9bb51d       292
87   air_2195cd5025a98033       253
88   air_21f5052d5330528d       288
89   air_220cba70c890b119       284
90   air_22682e965418936f       273
91   air_228f10bec0bda9c8       239
92   air_229d7e508d9f1b5e       209
93   air_232dcee6f7c51d37       475
94   air_234d3dbf7f3d5a50       441
95   air_23e1b11aee2a1407       284
96   air_23ee674e91469086       462
97   air_24b9b2a020826ede       394
98   air_24e8414b9b07decb       234
99   air_2545dd3a00f265e2       248
100  air_256be208a979e023       278
101  air_2570ccb93badde68       408
102  air_258ad2619d7bff9a       267
103  air_258dc112912fc458       240
104  air_25c583983246b7b0       202
105  air_25d8e5cc57dd87d9       428
106  air_25e9888d30b386df       320
107  air_2634e41551e9807d       457
108  air_26c5bbeb7bb82bf1       376
109  air_26f10355d9b4d82a       244
110  air_2703dcb33192b181  

In [239]:
q8c = q8.merge(q8b,how="left",on="id")

In [240]:
q8c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 252108 entries, 0 to 252107
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                252108 non-null  object        
 1   visit_date        252108 non-null  datetime64[ns]
 2   visitors          252108 non-null  int64         
 3   calendar_date     252108 non-null  object        
 4   day_of_week       252108 non-null  object        
 5   holiday           252108 non-null  int64         
 6   genre             252108 non-null  object        
 7   area              252108 non-null  object        
 8   latitude          252108 non-null  float64       
 9   longitude         252108 non-null  float64       
 10  reserve_visitors  108394 non-null  float64       
 11  rn                252108 non-null  int64         
 12  max_rows          252108 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(4), object(5)
memor

In [243]:
(q8c['rn'] <=  q8c['max_rows'] -15)

0          True
1          True
2          True
3          True
4          True
          ...  
252103    False
252104    False
252105    False
252106    False
252107    False
Length: 252108, dtype: bool

In [249]:
?np.where

In [250]:
q8c['train_test'] = np.where((q8c['rn'] <=  q8c['max_rows'] -15),'train','test')

In [264]:
# q8c.iloc[48574]
q8c.iloc[208574]

id                              air_d1418d6fd6d634f2
visit_date                       2017-02-21 00:00:00
visitors                                          10
calendar_date                             2017-02-21
day_of_week                                  Tuesday
holiday                                            0
genre                                        Izakaya
area                Hyōgo-ken Kōbe-shi Motomachidōri
latitude                                     34.6882
longitude                                    135.187
reserve_visitors                                  16
rn                                               214
max_rows                                         274
train_test                                     train
Name: 208574, dtype: object

In [261]:
q8c[q8c['id']=='air_00a91d42b08b08d9']['train_test'].value_counts()

train    217
test      15
Name: train_test, dtype: int64

In [263]:
q8c[q8c['id']=='air_36429b5ca4407b3e']['train_test'].value_counts()

train    245
test      15
Name: train_test, dtype: int64

In [265]:
q8c[q8c['id']=='air_d1418d6fd6d634f2']['train_test'].value_counts()

train    259
test      15
Name: train_test, dtype: int64

In [279]:
q8t = (q8c.groupby(by=['id','train_test']).count()['visit_date']).reset_index()

In [280]:
q8t

id train_test  visit_date
0     air_00a91d42b08b08d9       test          15
1     air_00a91d42b08b08d9      train         217
2     air_0164b9927d20bcc3       test          15
3     air_0164b9927d20bcc3      train         134
4     air_0241aa3964b7f861       test          15
...                    ...        ...         ...
1653  air_fef9ccb3ba0da2f7      train         230
1654  air_ffcc2d5087e1b476       test          15
1655  air_ffcc2d5087e1b476      train         228
1656  air_fff68b929994bfbd       test          15
1657  air_fff68b929994bfbd      train         254

[1658 rows x 3 columns]

In [287]:
q8t.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1658 entries, 0 to 1657
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          1658 non-null   object
 1   train_test  1658 non-null   object
 2   visit_date  1658 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 39.0+ KB


In [294]:
((q8t[q8t['train_test']=='test']['visit_date']==15).reset_index())['visit_date'].value_counts()

True    829
Name: visit_date, dtype: int64

In [300]:
q8b.max_rows.min()
# All the restaurants had at least 20 observations.

20

### Grouping & Merging

In this section of the lab, we are going to create different types of summary statistics -- where the rows for an individual sample can be compared with a larger group statistic.

**Bonus:** If you want to make this a little bit more effective, instead of using the entire `df`, try using a grouping from the `train` variable you just created, and use the grouping's values to populate both the training and test sets.

Use the technique discussed in class to create columns for the following stats:

**Question 1:** Create columns that list the average, median and standard deviation of visitors for each restaurant

In [7]:
# your answer here

In [39]:
df.head()

id visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  year  quarter  
0               NaN  2016        1  
1               NaN  2016        1  
2               NaN  2016        1  
3               NaN  2016        1  
4               NaN  2016        1

In [41]:
restaurant_stats = df.groupby('id')['visitors'].agg([np.mean,np.median,np.std]).rename({'mean':'restaurant-mean','median':'restaurant-median','std':"restaurant-standard"},axis=1)

In [42]:
restaurant_stats.head()

restaurant-mean  restaurant-median  restaurant-standard
id                                                                           
air_00a91d42b08b08d9        26.081897               26.0            12.435364
air_0164b9927d20bcc3         9.248322                8.0             6.348980
air_0241aa3964b7f861         9.896465                9.0             6.214877
air_0328696196e46f18         7.939655                6.0             6.733807
air_034a3d5b40d5b1b1        14.828685               12.0            13.154107

In [44]:
df.merge(restaurant_stats,left_on='id',right_index=True,how='left')

id visit_date  visitors calendar_date day_of_week  \
0       air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1       air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2       air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3       air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4       air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   
...                      ...        ...       ...           ...         ...   
252103  air_a17f0778617c76e2 2017-04-21        49    2017-04-21      Friday   
252104  air_a17f0778617c76e2 2017-04-22        60    2017-04-22    Saturday   
252105  air_a17f0778617c76e2 2017-03-26        69    2017-03-26      Sunday   
252106  air_a17f0778617c76e2 2017-03-20        31    2017-03-20      Monday   
252107  air_a17f0778617c76e2 2017-04-09        26    2017-04-09      Sunday   

        holiday           genre                          area   latitude  \
0             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
1             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
2             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
3             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
4             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
...         ...             ...                           ...        ...   
252103        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252104        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252105        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252106        1  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252107        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   

         longitude  reserve_visitors  year  quarter  restaurant-mean  \
0       139.751599               NaN  2016        1        22.782609   
1       139.751599               NaN  2016        1        22.782609   
2       139.751599               NaN  2016        1        22.782609   
3       139.751599               NaN  2016        1        22.782609   
4       139.751599               NaN  2016        1        22.782609   
...            ...               ...   ...      ...              ...   
252103  135.197852               6.0  2017        2        44.595745   
252104  135.197852              37.0  2017        2        44.595745   
252105  135.197852              35.0  2017        1        44.595745   
252106  135.197852               3.0  2017        1        44.595745   
252107  135.197852              32.0  2017        2        44.595745   

        restaurant-median  restaurant-standard  
0                    22.0            11.810526  
1                    22.0            11.810526  
2                    22.0            11.810526  
3                    22.0            11.810526  
4                    22.0            11.810526  
...                   ...                  ...  
252103               43.0            24.796265  
252104               43.0            24.796265  
252105               43.0            24.796265  
252106               43.0            24.796265  
252107               43.0            24.796265  

[252108 rows x 16 columns]

**Question 2:** Create a column that lists the average and median sales amount for each restaurant on a particular day of the week.

In [8]:
# your answer here

In [46]:
day_of_week_stats = df.groupby(['id','day_of_week'])['visitors'].agg([np.mean,np.median]).rename({'mean':'day_of_week-mean','median':'day_of_week-median'},axis=1)

In [49]:
day_of_week_stats.head(10)

day_of_week-mean  day_of_week-median
id                   day_of_week                                      
air_00a91d42b08b08d9 Friday              36.500000                35.5
                     Monday              22.457143                19.0
                     Saturday            14.973684                11.0
                     Sunday               2.000000                 2.0
                     Thursday            29.868421                30.0
                     Tuesday             24.350000                24.5
                     Wednesday           28.125000                28.0
air_0164b9927d20bcc3 Friday              11.464286                10.5
                     Monday               7.500000                 6.0
                     Saturday             6.409091                 4.5

In [47]:
df.merge(day_of_week_stats,left_on=['id','day_of_week'],right_index=True, how='left')

id visit_date  visitors calendar_date day_of_week  \
0       air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1       air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2       air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3       air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4       air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   
...                      ...        ...       ...           ...         ...   
252103  air_a17f0778617c76e2 2017-04-21        49    2017-04-21      Friday   
252104  air_a17f0778617c76e2 2017-04-22        60    2017-04-22    Saturday   
252105  air_a17f0778617c76e2 2017-03-26        69    2017-03-26      Sunday   
252106  air_a17f0778617c76e2 2017-03-20        31    2017-03-20      Monday   
252107  air_a17f0778617c76e2 2017-04-09        26    2017-04-09      Sunday   

        holiday           genre                          area   latitude  \
0             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
1             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
2             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
3             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
4             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
...         ...             ...                           ...        ...   
252103        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252104        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252105        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252106        1  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252107        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   

         longitude  reserve_visitors  year  quarter  day_of_week-mean  \
0       139.751599               NaN  2016        1         23.843750   
1       139.751599               NaN  2016        1         20.292308   
2       139.751599               NaN  2016        1         34.738462   
3       139.751599               NaN  2016        1         27.651515   
4       139.751599               NaN  2016        1         13.754386   
...            ...               ...   ...      ...               ...   
252103  135.197852               6.0  2017        2         68.428571   
252104  135.197852              37.0  2017        2         57.285714   
252105  135.197852              35.0  2017        1         44.000000   
252106  135.197852               3.0  2017        1         26.333333   
252107  135.197852              32.0  2017        2         44.000000   

        day_of_week-median  
0                     25.0  
1                     21.0  
2                     35.0  
3                     27.0  
4                     12.0  
...                    ...  
252103                65.0  
252104                60.0  
252105                34.5  
252106                27.5  
252107                34.5  

[252108 rows x 15 columns]

**Question 3:** Create columns that display the average and median sales amount for each genre in each city on each day of the week.  Create a column called `city` that captures the first value of `area` in order to this.  Values should be `Tokyo`, `Hiroshima`, etc.  **Hint:** You should use the `str` attribute combined with `split` in order to do this.

In [9]:
# your answer here

In [51]:
df.head(5)

id visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  year  quarter  
0               NaN  2016        1  
1               NaN  2016        1  
2               NaN  2016        1  
3               NaN  2016        1  
4               NaN  2016        1

In [56]:
df['city'] = df['area'].apply(lambda x: x.split(' ')[0])

In [59]:
df['city'].value_counts()

Tōkyō-to         133063
Fukuoka-ken       39645
Ōsaka-fu          22821
Hyōgo-ken         17846
Hokkaidō          13055
Hiroshima-ken      9858
Miyagi-ken         5959
Shizuoka-ken       5798
Niigata-ken        4063
Name: city, dtype: int64

In [62]:
city_genre_stats = df.groupby(['genre','city'])['visitors'].agg([np.mean,np.median]).rename({'mean':'genre_city-mean','median':'genre_city-median'},axis=1)

In [63]:
df.merge(city_genre_stats,left_on=['genre','city'],right_index=True, how='left')

id visit_date  visitors calendar_date day_of_week  \
0       air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1       air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2       air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3       air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4       air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   
...                      ...        ...       ...           ...         ...   
252103  air_a17f0778617c76e2 2017-04-21        49    2017-04-21      Friday   
252104  air_a17f0778617c76e2 2017-04-22        60    2017-04-22    Saturday   
252105  air_a17f0778617c76e2 2017-03-26        69    2017-03-26      Sunday   
252106  air_a17f0778617c76e2 2017-03-20        31    2017-03-20      Monday   
252107  air_a17f0778617c76e2 2017-04-09        26    2017-04-09      Sunday   

        holiday           genre                          area   latitude  \
0             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
1             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
2             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
3             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
4             0      Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068   
...         ...             ...                           ...        ...   
252103        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252104        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252105        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252106        1  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   
252107        0  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri  34.695124   

         longitude  reserve_visitors  year  quarter       city  \
0       139.751599               NaN  2016        1   Tōkyō-to   
1       139.751599               NaN  2016        1   Tōkyō-to   
2       139.751599               NaN  2016        1   Tōkyō-to   
3       139.751599               NaN  2016        1   Tōkyō-to   
4       139.751599               NaN  2016        1   Tōkyō-to   
...            ...               ...   ...      ...        ...   
252103  135.197852               6.0  2017        2  Hyōgo-ken   
252104  135.197852              37.0  2017        2  Hyōgo-ken   
252105  135.197852              35.0  2017        1  Hyōgo-ken   
252106  135.197852               3.0  2017        1  Hyōgo-ken   
252107  135.197852              32.0  2017        2  Hyōgo-ken   

        genre_city-mean  genre_city-median  
0             17.928582               14.0  
1             17.928582               14.0  
2             17.928582               14.0  
3             17.928582               14.0  
4             17.928582               14.0  
...                 ...                ...  
252103        21.655303               17.0  
252104        21.655303               17.0  
252105        21.655303               17.0  
252106        21.655303               17.0  
252107        21.655303               17.0  

[252108 rows x 16 columns]

In [64]:
df.head(5)

id visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   reserve_visitors  year  quarter      city  
0               NaN  2016        1  Tōkyō-to  
1               NaN  2016        1  Tōkyō-to  
2               NaN  2016        1  Tōkyō-to  
3               NaN  2016        1  Tōkyō-to  
4               NaN  2016        1  Tōkyō-to

In [66]:
df_all = df.merge(restaurant_stats,left_on='id',right_index=True,how='left')
df_all = df_all.merge(day_of_week_stats,left_on=['id','day_of_week'],right_index=True, how='left')
df_all = df_all.merge(city_genre_stats,left_on=['genre','city'],right_index=True, how='left')

In [67]:
df_all.head()

id visit_date  visitors calendar_date day_of_week  \
0  air_ba937bf13d40fb24 2016-01-13        25    2016-01-13   Wednesday   
1  air_ba937bf13d40fb24 2016-01-14        32    2016-01-14    Thursday   
2  air_ba937bf13d40fb24 2016-01-15        29    2016-01-15      Friday   
3  air_ba937bf13d40fb24 2016-01-16        22    2016-01-16    Saturday   
4  air_ba937bf13d40fb24 2016-01-18         6    2016-01-18      Monday   

   holiday       genre                          area   latitude   longitude  \
0        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
1        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
2        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
3        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   
4        0  Dining bar  Tōkyō-to Minato-ku Shibakōen  35.658068  139.751599   

   ...  year  quarter      city restaurant-mean  restaurant-median  \
0  ...  2016        1  Tōkyō-to       22.782609               22.0   
1  ...  2016        1  Tōkyō-to       22.782609               22.0   
2  ...  2016        1  Tōkyō-to       22.782609               22.0   
3  ...  2016        1  Tōkyō-to       22.782609               22.0   
4  ...  2016        1  Tōkyō-to       22.782609               22.0   

   restaurant-standard  day_of_week-mean  day_of_week-median  genre_city-mean  \
0            11.810526         23.843750                25.0        17.928582   
1            11.810526         20.292308                21.0        17.928582   
2            11.810526         34.738462                35.0        17.928582   
3            11.810526         27.651515                27.0        17.928582   
4            11.810526         13.754386                12.0        17.928582   

   genre_city-median  
0               14.0  
1               14.0  
2               14.0  
3               14.0  
4               14.0  

[5 rows x 21 columns]

In [68]:
df_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252108 entries, 0 to 252107
Data columns (total 21 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   id                   252108 non-null  object        
 1   visit_date           252108 non-null  datetime64[ns]
 2   visitors             252108 non-null  int64         
 3   calendar_date        252108 non-null  object        
 4   day_of_week          252108 non-null  object        
 5   holiday              252108 non-null  int64         
 6   genre                252108 non-null  object        
 7   area                 252108 non-null  object        
 8   latitude             252108 non-null  float64       
 9   longitude            252108 non-null  float64       
 10  reserve_visitors     108394 non-null  float64       
 11  year                 252108 non-null  int64         
 12  quarter              252108 non-null  int64         
 13  city          